src/editable_tree/mod.rs

```diff
@@ -107,59 +200,90 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
 
     /// Move the cursor in a given direction across the tree.  Returns [`Some`] error string if an
     /// error is found, or [`None`] if the movement was possible.
-    pub fn move_cursor(&mut self, direction: Direction) -> Option<String> {
+    pub fn move_cursor(&mut self, direction: Direction) -> EditResult {
         let (current_cursor, cursor_parent) = self.cursor_and_parent();
         match direction {
             Direction::Down => {
                 if current_cursor.children().is_empty() {
-                    Some("Cannot move down the tree if the cursor has no children.".to_string())
-                } else {
-                    self.current_cursor_path.push(0);
-                    None
+                    return EditResult::Err(EditErr::MoveToNonexistentChild);
                 }
+                self.current_cursor_path.push(0);
             }
             Direction::Up => {
                 if self.current_cursor_path.is_root() {
-                    return Some("Cannot move to the parent of the root.".to_string());
+                    return EditResult::Err(EditErr::MoveToParentOfRoot);
                 }
                 self.current_cursor_path.pop();
-                None
             }
             Direction::Prev => {
-                if let Some(index) = self.current_cursor_path.last_mut() {
-                    if *index == 0 {
-                        Some("Cannot move before the first child of a node.".to_string())
-                    } else {
-                        *index -= 1;
-                        None
-                    }
-                } else {
-                    Some("Cannot move to a sibling of the root.".to_string())
+                let index = match self.current_cursor_path.last_mut() {
+                    Some(index) => index,
+                    None => return EditResult::Err(EditErr::MoveToSiblingOfRoot),
+                };
```

Comment:
This whole thing (also repeated elsewhere) can be simplified like this (assuming `EditResult` becomes type alias of `Result`):

```rust
let index = self.current_cursor_path.last_mut().ok_or(EditErr::MoveToSiblingOfRoot)?;
```

Reply:
Fixed